onnxruntime/core/util/thread_utils.cc

```diff
@@ -7,9 +7,109 @@
 #endif
 #include <thread>
 #include "core/session/ort_apis.h"
+#include "core/common/logging/logging.h"
+
 
 namespace onnxruntime {
+
+#ifdef _WIN32
+ThreadAffinities GetDefaultThreadAffinities() {
+  DWORD returnLength = 0;
+  GetLogicalProcessorInformationEx(RelationGroup, nullptr, &returnLength);
+  auto last_error = GetLastError();
+  if (last_error != ERROR_INSUFFICIENT_BUFFER) {
+    LOGS_DEFAULT(ERROR) << "GetLogicalProcessorInformationEx failed to obtain buffer length. error code: "
+                        << last_error
+                        << " error msg: " << std::system_category().message(last_error);
+    return {};
+  }
+
+  std::unique_ptr<char[]> allocation = std::make_unique<char[]>(returnLength);
+  SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX* processorInfos = reinterpret_cast<SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX*>(allocation.get());
+
+  if (!GetLogicalProcessorInformationEx(RelationGroup, processorInfos, &returnLength)) {
+    last_error = GetLastError();
+    LOGS_DEFAULT(ERROR) << "GetLogicalProcessorInformationEx failed to obtain processor info. error code: "
+                        << last_error
+                        << " error msg: " << std::system_category().message(last_error);
+    return {};
+  }
+
+  WORD group_id = 0;
+  static constexpr WORD max_group_id = std::numeric_limits<WORD>::max();
+  ThreadAffinities thread_affinities;
+  // size of KAFFINITY varies, must not exceed boundary
+  int64_t num_bit_in_affinity_ = sizeof(KAFFINITY) << 3;
+  // allow only one thread for every two logical processors
+  int64_t max_num_thread_per_group = num_bit_in_affinity_ >> 1;
+
+  auto num_processor_info = returnLength / sizeof(SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX);
+  for (int64_t i = 0; i < static_cast<int64_t>(num_processor_info); ++i) {
+    if (processorInfos[i].Relationship != RelationGroup) {
+      LOGS_DEFAULT(ERROR) << "Returned processors not belong to same group";
+      return {};
+    }
+    for (int64_t j = 0; j < static_cast<int>(processorInfos[i].Group.ActiveGroupCount); ++j) {
+      const auto& groupInfo = processorInfos[i].Group.GroupInfo[j];
+      LOGS_DEFAULT(INFO) << "Discovered process group: " << group_id << ", active processor mask: " << groupInfo.ActiveProcessorMask;
+      KAFFINITY thread_affinity = 3UL;
+      // allow only one thread for every two logical processors
+      int64_t num_of_physical_cores = groupInfo.ActiveProcessorCount >> 1;
+      int64_t max_num_thread_cur_group = std::min( max_num_thread_per_group, num_of_physical_cores);
+      for (int64_t k = 0; k < max_num_thread_cur_group; ++k) {
+        thread_affinities.push_back({static_cast<int64_t>(group_id), static_cast<int64_t>(thread_affinity)});
+        thread_affinity <<= 2;
+      }
+      if (group_id == max_group_id) {
+        // preventing overflow
+        return std::move(thread_affinities);
```

Comment:
It's a pair of integers, so is this std::move required? In any case, won't the return value optimization in the compiler handle it?

Reply:
They are pairs of integers returned, not a pair of integer. And compiler(s) are not universally committed to do a move on a named local var since it is after all not a right value at all.

Reply:
Ah, it's a vector of thread_affinities. I missed that.

Reply:
> They are pairs of integers returned, not a pair of integer. And compiler(s) are not universally committed to do a move on a named local var since it is after all not a right value at all.

They are now committed.

Reply:
removed, thanks!